modules/gapi/src/compiler/passes/pattern_matching.cpp

```diff
@@ -0,0 +1,371 @@
+            }
+        }
+
+        ++i;
+    }
+
+    std::unordered_map<ade::NodeHandle, ade::NodeHandle, ade::HandleHasher<ade::Node>> inputApiMatch(firstPatternDataNodes.size());
+    std::unordered_map<ade::NodeHandle, ade::NodeHandle, ade::HandleHasher<ade::Node>> outputApiMatch(lastPatternDataNodes.size());
+
+    bool matched = true;
+
+    if (!notFound) {
+        VisitedMatchings matchedVisitedFirstDataNodes;
+        for (auto it = subgraphIns.begin(); it != subgraphIns.end() && matched; ++it) {
+            auto match = *it;
+            auto patternInputEdges = match.first->inEdges();
+            auto compInputEdges = match.second->inEdges();
+
+            if (match.first->inNodes().size() < match.second->inNodes().size()) {
+                matchedVisitedFirstDataNodes.clear();
+                inputApiMatch.clear();
+                matched = false;
+                break;
+            }
+
+            for (auto patternIt = patternInputEdges.begin(); patternIt != patternInputEdges.end(); ++patternIt) {
+
+                if ((*patternIt)->srcNode()->inEdges().size() != 0) {
+                    continue;
+                }
+
+                auto matchedIt = std::find_if(compInputEdges.begin(), compInputEdges.end(),
+                    [&patternIt, &patternGraph, &compGraph, &matchedVisitedFirstDataNodes](const ade::EdgeHandle& compEdge) -> bool {
+                    auto patternInputPort = patternGraph.metadata(*patternIt).get<cv::gimpl::Input>().port;
+                    auto compInputPort = compGraph.metadata(compEdge).get<cv::gimpl::Input>().port;
+
+                    if (patternInputPort != compInputPort) {
+                        return false;
+                    }
+
+                    auto foundit = std::find_if(matchedVisitedFirstDataNodes.begin(), matchedVisitedFirstDataNodes.end(), [&patternIt](std::pair<ade::NodeHandle, ade::NodeHandle> matchedNodes) { return (*patternIt)->srcNode() == matchedNodes.first; });
+                    if (foundit != matchedVisitedFirstDataNodes.end()) {
+                        if (compEdge->srcNode() != foundit->second) {
+                            return false;
+                        }
+                    }
+
+                    //shall be map in this case as doesn't require iterations during modification
+                    matchedVisitedFirstDataNodes.push_back({ (*patternIt)->srcNode(), compEdge->srcNode() });
+
+                    return true;
+                });
+
+                if (matchedIt == compInputEdges.end()) {
+                    matchedVisitedFirstDataNodes.clear();
+                    inputApiMatch.clear();
+                    matched = false;
+                    break;
+                }
+                inputApiMatch[(*patternIt)->srcNode()] = (*matchedIt)->srcNode();
+            }
+
+        }
+
+        if (matched) {
+            std::unordered_set<ade::NodeHandle, ade::HandleHasher<ade::Node>> visitedLastDataNodes;
```

Comment:
Not sure exactly what is going on, but can't we somehow memorize those nodes when matching?

Reply:
Here is matching for last data nodes. It will be performed after successful matching of pattern graph w/o first and last data nodes and successful matching of first data nodes.

This set was intent to check already visited last data nodes during walking through previous level of OP nodes and their descendants (bfs bypass). But for data nodes it is impossible to be already visited as it may mean, that two op nodes put their result into one allocated resource. So this set will be removed (after some more thinking) together with the code at 334 line.

Reply:
> Not sure exactly what is going on, but can't we somehow memorize those nodes when matching?

Here is matching for last data nodes. It will be performed after successful matching of pattern graph w/o first and last data nodes and successful matching of first data nodes.
This set was intent to check already visited last data nodes during walking through previous level of OP nodes and their descendants (bfs bypass). But for data nodes it is impossible to be already visited as it may mean, that two op nodes put their result into one allocated resource. So this set will be removed (after some more thinking) together with the code at 334 line.

Reply:
Got it, thanks!

Reply:
Code is removed.